stl/inc/xmemory

```diff
@@ -866,28 +868,22 @@ public:
     static constexpr size_t _Minimum_allocation_alignment = _Asan_granularity;
 };
 
+#if _HAS_DEPRECATED_ALLOCATOR_VOID || _HAS_DEPRECATED_ALLOCATOR_MEMBERS
 template <>
-class allocator<void> {
+class _CXX17_DEPRECATE_ALLOCATOR_VOID allocator<void> {
 public:
     using value_type = void;
 #if _HAS_DEPRECATED_ALLOCATOR_MEMBERS
     _CXX17_DEPRECATE_OLD_ALLOCATOR_MEMBERS typedef void* pointer;
     _CXX17_DEPRECATE_OLD_ALLOCATOR_MEMBERS typedef const void* const_pointer;
-#endif // _HAS_DEPRECATED_ALLOCATOR_MEMBERS
-
-    using size_type       = size_t;
-    using difference_type = ptrdiff_t;
 
-    using propagate_on_container_move_assignment           = true_type;
-    using is_always_equal _CXX20_DEPRECATE_IS_ALWAYS_EQUAL = true_type;

```

Comment:
If we allow users to override `_HAS_DEPRECATED_ALLOCATOR_VOID` or `_HAS_DEPRECATED_ALLOCATOR_MEMBERS`, it seems to me that these members should be present when `_HAS_CXX20` to facilitate mixing such "broken" code with code that expects a conformant C++20 `allocator<void>`. If we _don't_ allow users to override those macros, their definitions in `<yvals_core.h>` should be unconditional.

Reply:
Should we also add `allocate`, `deallocate`, and `allocate_at_least` if `_HAS_CXX20` (or `_HAS_CXX23`)?

If `allocator<void>` is a separated full specialization, I think it is possible to make `allocator<void>::allocate` and `allocator<void>::allocate_at_least` ill-formed on instantiation by making them templates.

Note that `allocator<void>::deallocate` is well-formed, but calling it always has UB.

Reply:
I'm adding these type when `_HAS_CXX20` now. I suspect that it's not good to add member functions (https://github.com/microsoft/STL/issues/527#issuecomment-591662259)...

Reply:
I agree that adding the member functions seems weird. Our goal here is to keep old code working without breaking new code, not to try to make a perfect facsimile that a determined opponent cannot distinguish from C++17 `allocator<void>`.